src/components/character-controller.js

```diff
@@ -0,0 +1,300 @@
+    eventSrc.removeEventListener("action_move_right", this.onMoveRight);
+    eventSrc.removeEventListener(
+      "action_dont_move_right",
+      this.onDontMoveRight
+    );
+    eventSrc.removeEventListener("rotateY", this.onRotateY);
+    eventSrc.removeEventListener(
+      "action_snap_rotate_left",
+      this.onSnapRotateLeft
+    );
+    eventSrc.removeEventListener(
+      "action_snap_rotate_right",
+      this.onSnapRotateRight
+    );
+  },
+
+  onStopMoving: function(event) {
+    this.accelerationInput.set(0, 0, 0);
+  },
+
+  onTranslateX: function(event) {
+    // The type check here is because the last axismove event that is captured here sends
+    // the el as the event.detail. This should probably be caught earlier.
+    this.accelerationInput.setX(
+      typeof event.detail === "number" ? event.detail : 0
+    );
+  },
+
+  onTranslateY: function(event) {
+    this.accelerationInput.setY(
+      typeof event.detail === "number" ? event.detail : 0
+    );
+  },
+
+  onTranslateZ: function(event) {
+    this.accelerationInput.setZ(
+      typeof event.detail === "number" ? event.detail : 0
+    );
+  },
+
+  onMoveForward: function(event) {
+    this.accelerationInput.z = 0.8;
+  },
+
+  onDontMoveForward: function(event) {
+    this.accelerationInput.z = 0;
+  },
+
+  onMoveBackward: function(event) {
+    this.accelerationInput.z = -0.8;
+  },
+
+  onDontMoveBackward: function(event) {
+    this.accelerationInput.z = 0;
+  },
+
+  onMoveLeft: function(event) {
+    this.accelerationInput.x = -0.8;
+  },
+
+  onDontMoveLeft: function(event) {
+    this.accelerationInput.x = 0;
+  },
+
+  onMoveRight: function(event) {
+    this.accelerationInput.x = 0.8;
+  },
+
+  onDontMoveRight: function(event) {
+    this.accelerationInput.x = 0;
+  },
+
+  onRotateY: function(event) {
+    if (typeof event.detail === "number") {
+      this.angularVelocity = event.detail;
+    } else {
+      this.angularVelocity = 0;
+    }
+  },
+
+  onSnapRotateLeft: function(event) {
+    this.pendingSnapRotationMatrix.copy(this.leftRotationMatrix);
+  },
+
+  onSnapRotateRight: function(event) {
+    this.pendingSnapRotationMatrix.copy(this.rightRotationMatrix);
+  },
+
+  onBoost: function(event) {
+    this.boost = event.detail;
+  },
+
+  tick: (function() {
+    var move = new THREE.Matrix4();
+    var trans = new THREE.Matrix4();
+    var transInv = new THREE.Matrix4();
+    var pivotPos = new THREE.Vector3();
+    var rotationAxis = new THREE.Vector3(0, 1, 0);
+    var yawMatrix = new THREE.Matrix4();
+    var rotationMatrix = new THREE.Matrix4();
+    var rotationInvMatrix = new THREE.Matrix4();
+    var pivotRotationMatrix = new THREE.Matrix4();
+    var pivotRotationInvMatrix = new THREE.Matrix4();
+    var position = new THREE.Vector3();
+    var currentPosition = new THREE.Vector3();
+    var movementVector = new THREE.Vector3();
+
+    return function(t, dt) {
+      const deltaSeconds = dt / 1000;
+      const rotationSpeed = -3;
```

Comment:
Probably best to make `rotationSpeed` a constant as well. Maybe put all of these in the schema with default values so they are configurable on the component.

Reply:
👍